src/Translation.jl

```diff
@@ -0,0 +1,338 @@
+"""
++(X::LazySet, v::AbstractVector) = Translation(X, v)
+
+"""
+    ⊕(X::LazySet, v::AbstractVector)
+
+Unicode alias constructor ⊕ (`oplus`) for the lazy translation operator.
+"""
+⊕(X::LazySet, v::AbstractVector) = Translation(X, v)
+
+# ============================ 
+# LazySet interface functions
+# ============================
+
+"""
+    dim(tr::Translation)::Int
+
+Return the dimension of a translation.
+
+### Input
+
+- `tr` -- translation
+
+### Output
+
+The dimension of a translation.
+"""
+function dim(tr::Translation)::Int
+    return length(tr.v)
+end
+
+
+"""
+    σ(d::AbstractVector{N}, tr::Translation{N}) where {N<:Real}
+
+Return the support vector of a translation.
+
+### Input
+
+- `d`  -- direction
+- `tr` -- translation
+
+### Output
+
+The support vector in the given direction.
+If the direction has norm zero, the result depends on the wrapped set.
+"""
+function σ(d::AbstractVector{N}, tr::Translation{N}) where {N<:Real}
+    return tr.v + σ(d, tr.X)
+end
+
+"""
+    ρ(d::AbstractVector{N}, tr::Translation{N}) where {N<:Real}
+
+Return the support function of a translation.
+
+### Input
+
+- `d`  -- direction
+- `tr` -- translation
+
+### Output
+
+The support function in the given direction.
+"""
+function ρ(d::AbstractVector{N}, tr::Translation{N}) where {N<:Real}
+    return ρ(d, Singleton(tr.v)) + ρ(d, tr.X)
+end
+
+"""
+    LinearMap(M::AbstractMatrix{N}, tr::Translation{N}) where {N<:Real}
+
+Return the lazy linear map of a translation.
+
+### Input
+
+- `M`  -- matrix
+- `tr` -- translation
+
+### Output
+
+The translation defined by the linear map.
+
+### Notes
+
+This method defines the simplification rule: ``M * (X⊕v) = (M*X) ⊕ (M*v)``,
+returning a new translation.
+"""
+function LinearMap(M::AbstractMatrix{N}, tr::Translation{N}) where {N<:Real}
+    return Translation(M * tr.X, M * tr.v)
+end
+
+"""
+    an_element(tr::Translation)
+
+Return some element of a translation.
+
+### Input
+
+- `tr` -- translation
+
+### Output
+
+An element in the translation.
+
+### Notes
+
+This function first asks for `an_element` function of the wrapped set, then
+translates this element according to the given translation vector. 
+"""
+function an_element(tr::Translation)
+    return an_element(tr.X) + tr.v
+end
+
+"""
+    isempty(tr::Translation)::Bool
+
+Return if a translation is empty or not.
+
+### Input
+
+- `tr` -- translation
+
+### Output
+
+`true` iff the wrapped set is empty.
+"""
+function isempty(tr::Translation)::Bool
+    return isempty(tr.X)
+end
+
+"""
+    constraints_list(tr::Translation{N, VN<:AbstractVector{N},
+                     S<:AbstractPolyhedron{N}}) where {N<:Real}
+
+Return the list of constraints of the translation of a polyhedron.
+
+### Input
+
+- `tr` -- lazy translation of a polyhedron
+
+### Output
+
+The list of constraints of the translation.
+
+### Notes
+
+We assume that the set wrapped by the lazy translation `X` is a polyhedron, i.e.,
+it offers a method `constraints_list(⋅)`.
```

Comment:
Hm, this sounds like you wanted to omit the set type constraint `<:AbstractPolyhedron{N}`. This could actually be helpful for our favorite application `Reachability` :+1:

Reply:
How about a trait-based solution like:

```julia
function constraints_list(tr::Translation{N}, ::Val{true}) where {N<:Real}
    constraints_X = constraints_list(tr.X)
    constraints_TX = similar(constraints_X)
    @inbounds for (i, ci) in enumerate(constraints_X)
        constraints_TX[i] = HalfSpace(ci.a, ci.b + dot(ci.a, tr.v))
    end
    return constraints_TX
end

function constraints_list(tr::Translation{N}) where {N<:Real}
    if applicable(constraints_list, tr.X)
        return constraints_list(tr, Val(true))
    else
        return constraints_list(tr, Val(false))
    end
end

function constraints_list(tr::Translation{N}, ::Val{false}) where {N<:Real}
    error("this function requires that the `constraints_list` method is appliable")
end
```

Example:

```julia
julia> constraints_list(rand(BallInf)+rand(2))
4-element Array{HalfSpace{Float64},1}:
 HalfSpace{Float64}([1.0, 0.0], 1.9328477832060151)
 HalfSpace{Float64}([0.0, 1.0], 1.1484644378760902)
 HalfSpace{Float64}([-1.0, -0.0], 0.7971425949785014)
 HalfSpace{Float64}([-0.0, -1.0], 1.5815259403084265)

julia> constraints_list(rand(Ball2)+rand(2))
ERROR: this function requires that the `constraints_list` method is appliable
Stacktrace:
 [1] error(::String) at ./error.jl:33

 [2] constraints_list(::Translation{Float64,Array{Float64,1},Ball2{Float64}}, ::Val{false}) at /Users/forets/.julia/dev/LazySets/src/Translation.jl:348
 [3] constraints_list(::Translation{Float64,Array{Float64,1},Ball2{Float64}}) at /Users/forets/.julia/dev/LazySets/src/Translation.jl:343
 [4] top-level scope at none:0
```

Reply:
Generally yes, but maybe we should do this in a second step and for all lazy operations.
For now, you can add the first method (without the second argument) I would say.

And by the way:
```julia
if x
    return f(y, true)
else
    return f(y, false)
end
```
↓
```julia
return f(y, x)
```

Reply:
Okay, now that you already added it, we can keep it :+1: